### PR TITLE
Builds on #1391, Adding support for Razer Chroma ARGB Controller

### LINKED ISSUE #1357: Support for Razer Chroma Addressable (ARGB) LED Controller
### Device Info

* Name: Razer Chroma Addressable LED controller
* Product Number: RZ34-021406
* Model Number: 
* Razer Website: https://www.razer.com/gaming-pc-accessories/razer-chroma-addressable-rgb-controller/RZ34-02140600-R3M1

### Device Information

USB ID: 1532:0F1F (interfaces 0, 1, 2)

### Packet Captures

[ChromaAddressableController_SynapseStartup.zip](https://github.com/openrazer/openrazer/files/5624610/ChromaAddressableController_SynapseStartup.zip)

At around 12 seconds I click the reset button and it rescans the number of attached LEDs.  This appears to use standard 90-byte reports:
[ChromaAddressable_ResetLEDSizeAuto.zip](https://github.com/openrazer/openrazer/files/5625872/ChromaAddressable_ResetLEDSizeAuto.zip)

### HID Descriptor

```
adam@adam-linux-desktop:~/Downloads$ sudo usbhid-dump -d 1532:0f1f
001:005:002:DESCRIPTOR         1606852527.819101
 05 0C 09 01 A1 01 06 00 FF 09 02 15 00 25 01 75
 08 95 10 81 00 C0

001:005:001:DESCRIPTOR         1606852527.822074
 05 0C 09 01 A1 01 06 00 FF 09 02 15 00 25 01 75
 08 96 40 01 B1 01 C0

001:005:000:DESCRIPTOR         1606852527.825013
 05 0C 09 01 A1 01 06 00 FF 09 02 15 00 25 01 75
 08 95 5A B1 01 C0
```

Interface 0 - 90 bytes (standard Razer report, existing control seems to work here, usage page 0x0C and usage 0x01)

Interface 1 - 320 bytes (looks to be direct mode packets that can hold the full 80 LEDs per channel RGB data, also 0x0C 0x01)

Interface 2 - 16 bytes (not sure what this is for, also 0x0C 0x01)

### Synapse Screenshots

Supports auto-detection of LED count and enabled ports, but also supports manual LED sizing.  Looks to cap out at a maximum of 80 LEDs per channel.

image

Individual channel brightness options

image

I just ordered this controller and would like to get it added to OpenRazer/OpenRGB.  My guess is that it's going to have a command to resize each channel in addition to the modes/direct control messages.

### PR DESCRIPTION
**Pull request by brett-smith**
Builds on https://github.com/openrazer/openrazer/pull/1391, adding daemon support and more driver features for Razer Chroma ARGB Controller.

### CODE REVIEW

File: daemon/openrazer_daemon/dbus_services/dbus_methods/argb_controller.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+
+def _get_channel_brightness(self, channel):
+    driver_path = self.get_driver_path(channel + '_led_brightness')
+    with open(driver_path, 'r') as driver_file:
+        return float(driver_file.read().strip()) / (255.0 / 100.0)
+
+
+@endpoint('razer.device.lighting.channel1', 'getChannel1Brightness', out_sig='d')
+def get_channel1_brightness(self):
```
Comment (z3ntu) [inline daemon/openrazer_daemon/dbus_services/dbus_methods/argb_controller.py:11]: Why not a method `get_channel_brightness(self, channel)`? Then we don't need to copy paste each method 6 times

Comment (z3ntu) [inline daemon/openrazer_daemon/dbus_services/dbus_methods/argb_controller.py:11]: I really dislike having all these methods copy-pasted 6 times but I guess it fits into the current API design...

**Comment by z3ntu** [comment]
Also what's the status on this PR? Does the added functionality work properly?

**Comment by hobeone** [comment]
I built debian packages from this branch and it seems to work well.  I've been able to get the module loaded and control things with polychromatic.

**Comment by milessi** [comment]
Did build from this and Chroma ARGB controller was working perfectly. I have 3 Razer devices (keyboard, mouse, controller). When controller was working perfectly my Razer Blackwidow Elite (1532:006c) become unrecognized. So did recognize mouse and controller but not the keyboard. With the v3.3.0 situation is back and 1532:0f1f (controller) is unrecongnized.

File: daemon/openrazer_daemon/dbus_services/dbus_methods/argb_controller.py (hunk 1 above)
Comment (z3ntu) [reply daemon/openrazer_daemon/dbus_services/dbus_methods/argb_controller.py:11]: @brett-smith Let's deviate here a bit from the current API design and just make a `getChannelBrightness` with parameter for channel ID. We also need a "nr. of channels" parameter somewhere that can be used. But I really really don't want to have 6 copies of these methods (even if it makes using this part of the API a bit more annoying as it's different)

**Comment by hobeone** [comment]
Reviving this thread - What needs to happen to get this merged?

**Comment by rhenndle** [comment]
I also have this controller and have successfully built this driver. Works a treat. It's basic and not as cool as the razer software, but it is more than functional.

**Comment by hobeone** [comment]
I had a second to sync this change with the main repo.  I created my own fork here https://github.com/hobeone/openrazer

Would it make sense to open a new pull request or somehow merge into this one?

@milessi would you mind trying a build from my fork to see if it resolves the recognition issue you saw?

**Comment by z3ntu** [comment]
@hobeone If you can fix the comments I added above, feel free to fix that and open a new PR.

**Comment by hobeone** [comment]
So something like this?

```
@endpoint('razer.device.lighting.channel', 'getChannelBrightness', in_sig='i', out_sig='d')
def get_channel_brightness(self, channel):
    """
    Get the channel brightness

    :param channel: Channel number to get the brightness of
    :type channel: int

    :return: Brightness
    :rtype: float
    """
    channel_name = 'channel{}'.format(channel)
    self.logger.debug("DBus call get_{}_brightness".format(channel_name))
    return _get_channel_brightness(self, channel_name)
```

**Comment by hobeone** [comment]
Reworked and set an initial pull request for review in #1929

**Comment by z3ntu** [comment]
Superseded by #1929

**Pull request closed without merging**